plutus-contract/src/Plutus/Contract/Test/ContractModel.hs

```diff
@@ -0,0 +1,1015 @@
+    --   `Test.QuickCheck.classify`, and `Test.QuickCheck.tabulate`. This function is called by
+    --   `propRunActions` (and friends) for any actions in the given `Actions`.
+    --
+    --   Statistics on which actions are executed are always collected.
+    monitoring :: (ModelState state, ModelState state)  -- ^ Model state before and after the action
+               -> Action state                          -- ^ The action that was performed
+               -> Property -> Property
+    monitoring _ _ = id
+
+    -- | In some scenarios it's useful to have actions that are never generated randomly, but only
+    --   used explicitly in `DL` script `action`s. To avoid these actions matching an `anyAction`
+    --   when shrinking, they can be marked `restricted`.
+    restricted :: Action state -> Bool
+    restricted _ = False
+
+-- | Model state lens
+makeLensesFor [("_contractState", "contractStateL")] 'ModelState
+makeLensesFor [("_currentSlot",   "currentSlotL")]   'ModelState
+makeLensesFor [("_lastSlot",      "lastSlotL")]      'ModelState
+makeLensesFor [("_balances",      "balancesL")]      'ModelState
+makeLensesFor [("_forged",        "forgedL")]        'ModelState
+
+-- | Get the contract-specific part of the model state.
+--
+--   `Spec` monad update functions: `$=` and `$~`.
+contractState :: Getter (ModelState state) state
+contractState = contractStateL
+
+-- | Get the current slot.
+--
+--   `Spec` monad update functions: `wait` and `waitUntil`.
+currentSlot :: Getter (ModelState state) Slot
+currentSlot = currentSlotL
+
+-- | Get the current wallet balances. These are delta balances, so they start out at zero and can be
+--   negative. The absolute balances used by the emulator can be set in the `CheckOptions` argument
+--   to `propRunActionsWithOptions`.
+--
+--   `Spec` monad update functions: `withdraw`, `deposit`, `transfer`.
+balances :: Getter (ModelState state) (Map Wallet Value)
+balances = balancesL
+
+-- | Get the current balance for a wallet. This is the delta balance, so it starts out at zero and
+--   can be negative. The absolute balance used by the emulator can be set in the `CheckOptions`
+--   argument to `propRunActionsWithOptions`.
+--
+--   `Spec` monad update functions: `withdraw`, `deposit`, `transfer`.
+balance :: Wallet -> Getter (ModelState state) Value
+balance w = balancesL . at w . non mempty
+
+-- | Get the amount of tokens forged so far. This is used to compute `lockedValue`.
+--
+--   `Spec` monad update functions: `forge` and `burn`.
+forged :: Getter (ModelState state) Value
+forged = forgedL
+
+-- | How much value is currently locked by contracts. This computed by subtracting the wallet
+--   `balances` from the `forged` value.
+lockedValue :: ModelState s -> Value
+lockedValue s = s ^. forged <> inv (fold $ s ^. balances)
+
+-- | Monads with read access to the model state: the `Spec` monad used in `nextState`, and the `DL`
```

Comment:
I guess maybe not `DL`, it's not clear to me if it lets you do `set`.

Reply:
Both could be `MonadState (ModelState s)`, but I don't want to expose that fact to the user. For `DL` letting the user change the state would just break things. For `Spec` it's not really a problem, but I think it's less error-prone going through the more fine-grained API.

Reply:
Yeah, I guess there's just a cognitive overhead for users. As a user, I have to look at the definition of this typeclass, try to understand why it's not `MonadState`, and remember a whole bunch of functions that I have to use instead of the ones I know from `MonadState`. 

I don't know what the balance of this tradeoff is!

Reply:
I'm making `Spec` a `MonadState`, but this doesn't make sense for `DL` so I'm keeping the type class.